File: cache/cache-annotation-processor/src/test/java/io/koraframework/cache/annotation/processor/AsyncCacheAopTests.java
package io.koraframework.cache.annotation.processor;

import io.koraframework.aop.annotation.processor.AopAnnotationProcessor;
import io.koraframework.cache.caffeine.CaffeineCacheModule;
import io.koraframework.cache.redis.RedisCacheClient;
import io.koraframework.cache.redis.RedisCacheModule;
import org.junit.jupiter.api.Test;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.function.BooleanSupplier;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

public class AsyncCacheAopTests extends AbstractCacheAnnotationProcessorTests implements CaffeineCacheModule, RedisCacheModule {

    @Test
    public void cacheableAsyncWithRedis() {
        compileRedis("""
            @Cacheable(value = DummyCache.class, mode = CacheMode.ASYNC)
            public String getValue(String arg1) {
                return value;
            }
            """);

        var client = new BlockingRedisCacheClient();
        var cache = newRedisCache(client);
        var service = newObject("$CacheableSync__AopProxy", cache, Executors.newSingleThreadExecutor());
        client.blockSet();

        assertThat(call(service, "getValue", "1")).isEqualTo("1");
        client.awaitBlocked();
        assertThat(call(cache, "get", "1")).isNull();

        client.release();
        assertEventually(() -> "1".equals(call(cache, "get", "1")));
    }

    @Test
    public void cachePutAsyncWithRedis() {
        compileRedis("""
            @CachePut(value = DummyCache.class, args = {"arg1"}, mode = CacheMode.ASYNC)
            public String putValue(String arg1) {
                return value;
            }
            """);

        var client = new BlockingRedisCacheClient();
        var cache = newRedisCache(client);
        var service = newObject("$CacheableSync__AopProxy", cache, Executors.newSingleThreadExecutor());
        client.blockSet();

        assertThat(call(service, "putValue", "1")).isEqualTo("1");
        client.awaitBlocked();
        assertThat(call(cache, "get", "1")).isNull();

        client.release();
        assertEventually(() -> "1".equals(call(cache, "get", "1")));
    }

    @Test
    public void cacheInvalidateAsyncWithRedis() {
        compileRedis("""
            @CacheInvalidate(value = DummyCache.class, mode = CacheMode.ASYNC)
            public void evictValue(String arg1) {
            }
            """);

        var client = new BlockingRedisCacheClient();
        var cache = newRedisCache(client);
        call(cache, "put", "1", "1");
        var service = newObject("$CacheableSync__AopProxy", cache, Executors.newSingleThreadExecutor());
        client.blockDelOne();

        call(service, "evictValue", "1");
        client.awaitBlocked();
        assertThat(call(cache, "get", "1")).isEqualTo("1");

        client.release();
        assertEventually(() -> call(cache, "get", "1") == null);
    }

    @Test
    public void cacheInvalidateAllAsyncWithRedis() {
        compileRedis("""
            @CacheInvalidateAll(value = DummyCache.class, mode = CacheMode.ASYNC)
            public void evictAll() {
            }
            """);

        var client = new BlockingRedisCacheClient();
        var cache = newRedisCache(client);
        call(cache, "put", "1", "1");
        call(cache, "put", "2", "2");
        var service = newObject("$CacheableSync__AopProxy", cache, Executors.newSingleThreadExecutor());
        client.blockDelMany();

        call(service, "evictAll");
        client.awaitBlocked();
        assertThat(call(cache, "get", "1")).isEqualTo("1");

        client.release();
        assertEventually(() -> call(cache, "get", "1") == null);
        assertEventually(() -> call(cache, "get", "2") == null);
    }

    @Test
    public void cacheableAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @Cacheable(value = DummyCache.class, mode = CacheMode.ASYNC)
            public String getValue(String arg1) {
                return value;
            }
            """);

        var cache = newCaffeineCache();
        var service = newObject("$CacheableSync__AopProxy", cache);

        assertThat(call(service, "getValue", "1")).isEqualTo("1");

        assertThat(call(cache, "get", "1")).isEqualTo("1");
    }

    @Test
    public void cachePutAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @CachePut(value = DummyCache.class, args = {"arg1"}, mode = CacheMode.ASYNC)
            public String putValue(String arg1) {
                return value;
            }
            """);

        var cache = newCaffeineCache();
        var service = newObject("$CacheableSync__AopProxy", cache);

        assertThat(call(service, "putValue", "1")).isEqualTo("1");

        assertThat(call(cache, "get", "1")).isEqualTo("1");
    }

    @Test
    public void cacheInvalidateAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @CacheInvalidate(value = DummyCache.class, mode = CacheMode.ASYNC)
            public void evictValue(String arg1) {
            }
            """);

        var cache = newCaffeineCache();
        call(cache, "put", "1", "1");
        var service = newObject("$CacheableSync__AopProxy", cache);

        call(service, "evictValue", "1");

        assertThat(call(cache, "get", "1")).isNull();
    }

    @Test
    public void cacheInvalidateAllAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @CacheInvalidateAll(value = DummyCache.class, mode = CacheMode.ASYNC)
            public void evictAll() {
            }
            """);

        var cache = newCaffeineCache();
        call(cache, "put", "1", "1");
        call(cache, "put", "2", "2");
        var service = newObject("$CacheableSync__AopProxy", cache);

        call(service, "evictAll");

        assertThat(call(cache, "get", "1")).isNull();
        assertThat(call(cache, "get", "2")).isNull();
    }

    private void compileRedis(String method) {
        var compileResult = compile(List.of(new CacheAnnotationProcessor(), new AopAnnotationProcessor()),
            """
                @Cache("dummy")
                public interface DummyCache extends io.koraframework.cache.redis.RedisCache<String, String> { }
                """, """
                public class CacheableSync {
                    public String value = "1";
                    
                    %s
                }
                """.formatted(method));
        compileResult.assertSuccess();
    }

    private void compileCaffeine(String method) {
        var compileResult = compile(List.of(new CacheAnnotationProcessor(), new AopAnnotationProcessor()),
            """
                @Cache("dummy")
                public interface DummyCache extends CaffeineCache<String, String> { }
                """, """
                public class CacheableSync {
                    public String value = "1";
                    
                    %s
                }
                """.formatted(method));
        compileResult.assertSuccess();
    }

    private Object newRedisCache(BlockingRedisCacheClient client) {
        return newObject("$DummyCache_Impl", CacheRunner.getRedisConfig(), client,
            defaultRedisCacheTelemetryFactory(null, null, null, null),
            stringRedisCacheKeyMapper(), stringRedisCacheValueMapper());
    }

    private Object newCaffeineCache() {
        return newObject("$DummyCache_Impl", CacheRunner.getCaffeineConfig(),
            caffeineCacheFactory(null), defaultCaffeineCacheTelemetryFactory(null, null, null, null));
    }

    @Override
    protected String commonImports() {
        return super.commonImports() +
               """
                   import io.koraframework.cache.annotation.CacheInvalidateAll;
                   import io.koraframework.cache.annotation.CacheMode;
                   """;
    }

    private static Object call(Object target, String method, Object... args) {
        try {
            for (var declaredMethod : target.getClass().getMethods()) {
                if (declaredMethod.getName().equals(method) && declaredMethod.getParameterCount() == args.length
                    && isAssignable(declaredMethod.getParameterTypes(), args)) {
                    return declaredMethod.invoke(target, args);
                }
            }
            throw new IllegalArgumentException("Method " + method + " wasn't found");
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static boolean isAssignable(Class<?>[] parameterTypes, Object[] args) {
        for (int i = 0; i < parameterTypes.length; i++) {
            if (args[i] != null && !parameterTypes[i].isAssignableFrom(args[i].getClass())) {
                return false;
            }
        }
        return true;
    }

    private static void assertEventually(BooleanSupplier assertion) {
        try {
            for (int i = 0; i < 50; i++) {
                if (assertion.getAsBoolean()) {
                    return;
                }
                Thread.sleep(100);
            }
            fail("Async cache operation didn't complete");
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            fail(e);
        }
    }

    private static final class BlockingRedisCacheClient implements RedisCacheClient {
        private final Map<ByteBuffer, ByteBuffer> cache = new ConcurrentHashMap<>();
        private final RedisCacheClient delegate = CacheRunner.lettuceClient(cache);
        private volatile Mode mode = Mode.NONE;
        private volatile CountDownLatch started = new CountDownLatch(0);
        private volatile CountDownLatch release = new CountDownLatch(0);

        void blockSet() {
            block(Mode.SET);
        }

        void blockDelOne() {
            block(Mode.DEL_ONE);
        }

        void blockDelMany() {
            block(Mode.DEL_MANY);
        }

        void awaitBlocked() {
            try {
                assertTrue(this.started.await(5, TimeUnit.SECONDS));
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                fail(e);
            }
        }

        void release() {
            this.release.countDown();
        }

        private void block(Mode mode) {
            this.mode = mode;
            this.started = new CountDownLatch(1);
            this.release = new CountDownLatch(1);
        }

        private void awaitRelease(Mode mode) {
            if (this.mode != mode) {
                return;
            }
            try {
                this.started.countDown();
                assertTrue(this.release.await(5, TimeUnit.SECONDS));
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                fail(e);
            }
        }

        @Override
        public List<byte[]> scan(byte[] prefix) {
            return delegate.scan(prefix);
        }

        @Override
        public byte[] get(byte[] key) {
            return delegate.get(key);
        }

        @Override
        public Map<byte[], byte[]> mget(byte[][] keys) {
            return delegate.mget(keys);
        }

        @Override
        public byte[] getex(byte[] key, long expireAfterMillis) {
            return delegate.getex(key, expireAfterMillis);
        }

        @Override
        public Map<byte[], byte[]> getex(byte[][] keys, long expireAfterMillis) {
            return delegate.getex(keys, expireAfterMillis);
        }

        @Override
        public void set(byte[] key, byte[] value) {
            awaitRelease(Mode.SET);
            delegate.set(key, value);
        }

        @Override
        public void mset(Map<byte[], byte[]> keyAndValue) {
            delegate.mset(keyAndValue);
        }

        @Override
        public void psetex(byte[] key, byte[] value, long expireAfterMillis) {
            awaitRelease(Mode.SET);
            delegate.psetex(key, value, expireAfterMillis);
        }

        @Override
        public void psetex(Map<byte[], byte[]> keyAndValue, long expireAfterMillis) {
            delegate.psetex(keyAndValue, expireAfterMillis);
        }

        @Override
        public long del(byte[] key) {
            awaitRelease(Mode.DEL_ONE);
            return delegate.del(key);
        }

        @Override
        public long del(byte[][] keys) {
            awaitRelease(Mode.DEL_MANY);
            return delegate.del(keys);
        }

        @Override
        public void flushAll() {
            delegate.flushAll();
        }

        private enum Mode {
            NONE,
            SET,
            DEL_ONE,
            DEL_MANY
        }
    }
}

File: cache/cache-symbol-processor/src/test/kotlin/io/koraframework/cache/symbol/processor/AsyncCacheAopTests.kt
package io.koraframework.cache.symbol.processor

import com.google.devtools.ksp.KspExperimental
import io.koraframework.aop.symbol.processor.AopSymbolProcessorProvider
import io.koraframework.cache.caffeine.CaffeineCacheModule
import io.koraframework.cache.redis.RedisCacheClient
import io.koraframework.cache.redis.RedisCacheModule
import io.koraframework.ksp.common.AbstractSymbolProcessorTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.nio.ByteBuffer
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

@KspExperimental
class AsyncCacheAopTests : AbstractSymbolProcessorTest(), CaffeineCacheModule, RedisCacheModule {

    override fun commonImports(): String {
        return super.commonImports() + """
            import io.koraframework.cache.annotation.Cache;
            import io.koraframework.cache.annotation.CacheInvalidate;
            import io.koraframework.cache.annotation.CacheInvalidateAll;
            import io.koraframework.cache.annotation.CacheMode;
            import io.koraframework.cache.annotation.CachePut;
            import io.koraframework.cache.annotation.Cacheable;
            import io.koraframework.cache.caffeine.CaffeineCache;
            import io.koraframework.cache.redis.RedisCache;
            
            """.trimIndent()
    }

    @Test
    fun cacheableAsyncWithRedis() {
        compileRedis("""
            @Cacheable(value = DummyCache::class, mode = CacheMode.ASYNC)
            open fun getValue(arg1: String): String {
                return value
            }
            """)

        val client = BlockingRedisCacheClient()
        val cache = newRedisCache(client)
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance, Executors.newSingleThreadExecutor())
        client.blockSet()

        assertEquals("1", service.call("getValue", "1"))
        client.awaitBlocked()
        assertNull(cache.call("get", "1"))

        client.release()
        assertEventually { "1" == cache.call("get", "1") }
    }

    @Test
    fun cachePutAsyncWithRedis() {
        compileRedis("""
            @CachePut(value = DummyCache::class, args = ["arg1"], mode = CacheMode.ASYNC)
            open fun putValue(arg1: String): String {
                return value
            }
            """)

        val client = BlockingRedisCacheClient()
        val cache = newRedisCache(client)
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance, Executors.newSingleThreadExecutor())
        client.blockSet()

        assertEquals("1", service.call("putValue", "1"))
        client.awaitBlocked()
        assertNull(cache.call("get", "1"))

        client.release()
        assertEventually { "1" == cache.call("get", "1") }
    }

    @Test
    fun cacheInvalidateAsyncWithRedis() {
        compileRedis("""
            @CacheInvalidate(value = DummyCache::class, mode = CacheMode.ASYNC)
            open fun evictValue(arg1: String) {
            }
            """)

        val client = BlockingRedisCacheClient()
        val cache = newRedisCache(client)
        cache.call("put", "1", "1")
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance, Executors.newSingleThreadExecutor())
        client.blockDelOne()

        service.call("evictValue", "1")
        client.awaitBlocked()
        assertEquals("1", cache.call("get", "1"))

        client.release()
        assertEventually { cache.call("get", "1") == null }
    }

    @Test
    fun cacheInvalidateAllAsyncWithRedis() {
        compileRedis("""
            @CacheInvalidateAll(value = DummyCache::class, mode = CacheMode.ASYNC)
            open fun evictAll() {
            }
            """)

        val client = BlockingRedisCacheClient()
        val cache = newRedisCache(client)
        cache.call("put", "1", "1")
        cache.call("put", "2", "2")
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance, Executors.newSingleThreadExecutor())
        client.blockDelMany()

        service.call("evictAll")
        client.awaitBlocked()
        assertEquals("1", cache.call("get", "1"))

        client.release()
        assertEventually { cache.call("get", "1") == null }
        assertEventually { cache.call("get", "2") == null }
    }

    @Test
    fun cacheableAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @Cacheable(value = DummyCache::class, mode = CacheMode.ASYNC)
            open fun getValue(arg1: String): String {
                return value
            }
            """)

        val cache = newCaffeineCache()
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance)

        assertEquals("1", service.call("getValue", "1"))

        assertEquals("1", cache.call("get", "1"))
    }

    @Test
    fun cachePutAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @CachePut(value = DummyCache::class, args = ["arg1"], mode = CacheMode.ASYNC)
            open fun putValue(arg1: String): String {
                return value
            }
            """)

        val cache = newCaffeineCache()
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance)

        assertEquals("1", service.call("putValue", "1"))

        assertEquals("1", cache.call("get", "1"))
    }

    @Test
    fun cacheInvalidateAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @CacheInvalidate(value = DummyCache::class, mode = CacheMode.ASYNC)
            open fun evictValue(arg1: String) {
            }
            """)

        val cache = newCaffeineCache()
        cache.call("put", "1", "1")
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance)

        service.call("evictValue", "1")

        assertNull(cache.call("get", "1"))
    }

    @Test
    fun cacheInvalidateAllAsyncWithCaffeineIsSync() {
        compileCaffeine("""
            @CacheInvalidateAll(value = DummyCache::class, mode = CacheMode.ASYNC)
            open fun evictAll() {
            }
            """)

        val cache = newCaffeineCache()
        cache.call("put", "1", "1")
        cache.call("put", "2", "2")
        val service = newObject("\$CacheableSync__AopProxy", cache.objectInstance)

        service.call("evictAll")

        assertNull(cache.call("get", "1"))
        assertNull(cache.call("get", "2"))
    }

    private fun compileRedis(method: String) {
        compile0(
            listOf(CacheSymbolProcessorProvider(), AopSymbolProcessorProvider()),
            """
            @Cache("dummy")
            interface DummyCache : RedisCache<String, String>
            """,
            """
            open class CacheableSync {
                var value = "1"
                
                $method
            }
            """
        ).assertSuccess()
    }

    private fun compileCaffeine(method: String) {
        compile0(
            listOf(CacheSymbolProcessorProvider(), AopSymbolProcessorProvider()),
            """
            @Cache("dummy")
            interface DummyCache : CaffeineCache<String, String>
            """,
            """
            open class CacheableSync {
                var value = "1"
                
                $method
            }
            """
        ).assertSuccess()
    }

    private fun newRedisCache(client: BlockingRedisCacheClient): TestObject {
        return newObject(
            "\$DummyCache_Impl",
            CacheRunner.getRedisConfig(),
            client,
            defaultRedisCacheTelemetryFactory(null, null, null, null),
            stringRedisCacheKeyMapper(),
            stringRedisCacheValueMapper()
        )
    }

    private fun newCaffeineCache(): TestObject {
        return newObject(
            "\$DummyCache_Impl",
            CacheRunner.getCaffeineConfig(),
            caffeineCacheFactory(null),
            defaultCaffeineCacheTelemetryFactory(null, null, null, null)
        )
    }

    private fun assertEventually(assertion: () -> Boolean) {
        repeat(50) {
            if (assertion()) {
                return
            }
            Thread.sleep(100)
        }
        fail<Nothing>("Async cache operation didn't complete")
    }

    private fun TestObject.call(method: String, vararg args: Any?): Any? {
        val javaMethod = objectInstance.javaClass.methods.firstOrNull {
            it.name == method && it.parameterCount == args.size && it.parameterTypes.isAssignable(args)
        } ?: throw IllegalArgumentException("Method $method wasn't found")
        return javaMethod.invoke(objectInstance, *args)
    }

    private fun Array<Class<*>>.isAssignable(args: Array<out Any?>): Boolean {
        for (i in indices) {
            if (args[i] != null && !this[i].isAssignableFrom(args[i]!!.javaClass)) {
                return false
            }
        }
        return true
    }

    private class BlockingRedisCacheClient : RedisCacheClient {
        private val cache = ConcurrentHashMap<ByteBuffer?, ByteBuffer?>()
        private val delegate = CacheRunner.lettuceClient(cache)
        private var mode = Mode.NONE
        private var started = CountDownLatch(0)
        private var release = CountDownLatch(0)

        fun blockSet() = block(Mode.SET)

        fun blockDelOne() = block(Mode.DEL_ONE)

        fun blockDelMany() = block(Mode.DEL_MANY)

        fun awaitBlocked() {
            assertTrue(started.await(5, TimeUnit.SECONDS))
        }

        fun release() {
            release.countDown()
        }

        private fun block(mode: Mode) {
            this.mode = mode
            started = CountDownLatch(1)
            release = CountDownLatch(1)
        }

        private fun awaitRelease(mode: Mode) {
            if (this.mode != mode) {
                return
            }
            started.countDown()
            assertTrue(release.await(5, TimeUnit.SECONDS))
        }

        override fun scan(prefix: ByteArray): List<ByteArray> = delegate.scan(prefix)

        override fun get(key: ByteArray): ByteArray? = delegate.get(key)

        override fun mget(keys: Array<ByteArray>): Map<ByteArray, ByteArray> = delegate.mget(keys)

        override fun getex(key: ByteArray, expireAfterMillis: Long): ByteArray? = delegate.getex(key, expireAfterMillis)

        override fun getex(keys: Array<ByteArray>, expireAfterMillis: Long): Map<ByteArray, ByteArray> = delegate.getex(keys, expireAfterMillis)

        override fun set(key: ByteArray, value: ByteArray) {
            awaitRelease(Mode.SET)
            delegate.set(key, value)
        }

        override fun mset(keyAndValue: MutableMap<ByteArray, ByteArray>) = delegate.mset(keyAndValue)

        override fun psetex(key: ByteArray, value: ByteArray, expireAfterMillis: Long) {
            awaitRelease(Mode.SET)
            delegate.psetex(key, value, expireAfterMillis)
        }

        override fun psetex(keyAndValue: MutableMap<ByteArray, ByteArray>, expireAfterMillis: Long) = delegate.psetex(keyAndValue, expireAfterMillis)

        override fun del(key: ByteArray): Long {
            awaitRelease(Mode.DEL_ONE)
            return delegate.del(key)
        }

        override fun del(keys: Array<ByteArray>): Long {
            awaitRelease(Mode.DEL_MANY)
            return delegate.del(keys)
        }

        override fun flushAll() = delegate.flushAll()

        private enum class Mode {
            NONE,
            SET,
            DEL_ONE,
            DEL_MANY
        }
    }
}
